### ISSUE TITLE
Sign error for Qy in multilayerqg.jl

### ISSUE THREAD

**Issue by mjclobo** [FourierFlows/GeophysicalFlows.jl#328]
This is an issue made to accompany my pull request  (#327 ) for this bug.

There is a sign error in the multilayerqg.jl calculation of Qy:
https://github.com/FourierFlows/GeophysicalFlows.jl/blob/c0c42cebd076b9cf828af1b956a08241a70febcc/src/multilayerqg.jl#LL367C1-L367C138

The first minus sign should be a plus sign. This creates erroneously large interior PV gradients.

The following figures can be reproduced using [this script](https://github.com/mjclobo/GeophysicalFlows.jl/blob/main/examples/matts_examples/GFjl_bug_plots.jl) and my forked version of GeophysicalFlows.jl. For each figure the left plot is of the meridional PV gradient and the right plot is of the vertical structure of the most unstable mode. The reduced gravity error mentioned here is brought up in discussion #325 . 

The first figure shows results when using density values of O(1), as in the [Phillips model example](https://fourierflows.github.io/GeophysicalFlowsDocumentation/stable/literated/multilayerqg_2layer/). The sign error leads to a large erroneous positive interior PV gradient and a deep Charney-type instability that arises from interactions between the negative bottom PV gradient and the positive interior PV gradient. The reduced gravity error creates an erroneous negative interior PV gradient, that then interacts with the positive surface PV gradient to manifest a surface-intensified Charney-like instability. Fixing both errors leads to no instability since the edges are too far apart (given the current background conditions) to interact and create an Eady-type instability.

![image](https://github.com/FourierFlows/GeophysicalFlows.jl/assets/104666665/cc0e8447-889e-45f9-bf78-fe888f04c761)

@wendazhang33, my office mate, pointed out that when you use realistic ocean density values, of O(1000), then the reduced gravity definition ceases to matter. In this case the orange line falls directly under the green line. Ironically, when O(rho)>>O(delta rho), the Boussinesq assumption applies, which means that a constant reference density should still be used in the denominator of the definition of reduced gravity. Regardless, the sign error is of the same order of consequence as when using O(1) density values, as is shown in the second plot, below.

![image](https://github.com/FourierFlows/GeophysicalFlows.jl/assets/104666665/c9dcbba8-54bb-4c25-a5f0-df9dcbe38fd7)

**Comment by navidcy** [comment]
thanks @mjclobo!!

Note, if you write your post like below (with an empty line between txt and link to code), the code will appear in the post.

```
this is some piece of code I'd like to attach

https://github.com/FourierFlows/GeophysicalFlows.jl/blob/c0c42cebd076b9cf828af1b956a08241a70febcc/src/multilayerqg.jl#L367-L367

I think this and that....
```

(You can edit your post and see.)

**Comment by navidcy** [comment]
Regarding the sign error:

Here's the code that computes `Qy` for each layer:

https://github.com/FourierFlows/GeophysicalFlows.jl/blob/c0c42cebd076b9cf828af1b956a08241a70febcc/src/multilayerqg.jl#L365-L369

In #327  you suggested changing

`- Fp[j] * (U[:, :, j+1] - U[:, :, j])` -> `+ Fp[j] * (U[:, :, j+1] - U[:, :, j])` for `j = 2,...,n-1`

But then this is inconsistent with layers `1` and `n`.

Do you agree with the equations in the [docs](https://fourierflows.github.io/GeophysicalFlowsDocumentation/stable/modules/multilayerqg/). If yes, then it seems to me that the sign error is on the `Fm` term in the interior, i.e.,

`+ Fm[j-1] * (U[:, :, j-1] - U[:, :, j]` -> `- Fm[j-1] * (U[:, :, j-1] - U[:, :, j]`

Right?

**Comment by navidcy** [comment]
> @Wendazhang33, my office mate, pointed out that when you use realistic ocean density values, of O(1000), then the reduced gravity definition ceases to matter. In this case the orange line falls directly under the green line. Ironically, when O(rho)>>O(delta rho), the Boussinesq assumption applies, which means that a constant reference density should still be used in the denominator of the definition of reduced gravity. Regardless, the sign error is of the same order of consequence as when using O(1) density values, as is shown in the second plot, below. Also note that with realistic ocean density values, the corrected code leads to a nice Eady instability, as expected.

Yes, I was realizing last night that we might be using the Boussinesq approximation under the hood in the buoyancy frequency definition. I'm trying to clear this up.

**Comment by mjclobo** [comment]
> Regarding the sign error:
> 
> Here's the code that computes `Qy` for each layer:
> 
> https://github.com/FourierFlows/GeophysicalFlows.jl/blob/c0c42cebd076b9cf828af1b956a08241a70febcc/src/multilayerqg.jl#L365-L369
> 
> In #327 you suggested changing
> 
> `- Fp[j] * (U[:, :, j+1] - U[:, :, j])` -> `+ Fp[j] * (U[:, :, j+1] - U[:, :, j])` for `j = 2,...,n-1`
> 
> But then this is inconsistent with layers `1` and `n`.
> 
> Do you agree with the equations in the [docs](https://fourierflows.github.io/GeophysicalFlowsDocumentation/stable/modules/multilayerqg/). If yes, then it seems to me that the sign error is on the `Fm` term in the interior, i.e.,
> 
> `+ Fm[j-1] * (U[:, :, j-1] - U[:, :, j]` -> `- Fm[j-1] * (U[:, :, j-1] - U[:, :, j]`
> 
> Right?

Shoot, you're right. Thanks for the catch! I closed the pull request. Is this the correct thing to do? Should I open a new one, or will you do it? Thanks for the help!

**Comment by navidcy** [comment]
Open a new one only with the sign error fix?

Let's leave the reference density discussion for a different PR.

**Comment by navidcy** [comment]
Try to keep the PR clean (i.e., don't include example scripts of your own in the PR as a suggestion to be merged in the main branch). You can always post code as part of a comment on the PR or here! :)

**Comment by mjclobo** [comment]
Thanks for the advice @navidcy. I  just opened PR #329.

Is it okay to have the interim commits in the pull request, since the actual difference between the codes ends up only being the one line? Or should I make sure to only have commits relevant to the issue shown? I couldn't figure out how to do the latter, but am happy to keep trying if it's preferable.

**Comment by navidcy** [comment]
That's alright :) When we merge we can do "Squash and merge" and that will compress all the commits into one.

**Comment by navidcy** [comment]
PR #329 dealt with the sign error!
